File: app/diophantine.py
"""Exact integer solving of linear Diophantine systems ``A x = b``.

The solver relies on the Smith normal form ``D = U @ A @ V`` where ``U``
and ``V`` are unimodular integer matrices (products of invertible integer
row / column operations).  ``A x = b`` has an integer solution iff the
Smith pivots divide the corresponding entries of the transformed target
``U @ b`` and the transformed target vanishes on every zero row of ``D``.

All arithmetic uses Python arbitrary-precision integers.  No floating
point conversion and no rounding of rationals occurs anywhere in this
module; divisions are exact Euclidean divisions whose remainders are
kept explicitly.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Sequence

Matrix = List[List[int]]


def _identity(size: int) -> Matrix:
    return [[1 if i == j else 0 for j in range(size)] for i in range(size)]


@dataclass
class SmithDecomposition:
    """Smith normal form data with ``D = U @ A @ V``."""

    d: Matrix  # m x n diagonal (Smith) matrix
    u: Matrix  # m x m unimodular: invertible integer row transforms
    v: Matrix  # n x n unimodular: invertible integer column transforms
    rank: int
    diagonal: List[int]  # positive pivots d_1 | d_2 | ... | d_rank


def smith_normal_form(matrix: Sequence[Sequence[int]]) -> SmithDecomposition:
    """Compute the Smith normal form of an integer matrix exactly."""
    a = [list(row) for row in matrix]
    m = len(a)
    n = len(a[0]) if m else 0
    u = _identity(m)
    v = _identity(n)

    def swap_rows(i: int, j: int) -> None:
        if i != j:
            a[i], a[j] = a[j], a[i]
            u[i], u[j] = u[j], u[i]

    def negate_row(i: int) -> None:
        a[i] = [-value for value in a[i]]
        u[i] = [-value for value in u[i]]

    def add_row_multiple(src: int, dst: int, factor: int) -> None:
        if factor:
            a[dst] = [x + factor * y for x, y in zip(a[dst], a[src])]
            u[dst] = [x + factor * y for x, y in zip(u[dst], u[src])]

    def swap_cols(i: int, j: int) -> None:
        if i != j:
            for row in a:
                row[i], row[j] = row[j], row[i]
            for row in v:
                row[i], row[j] = row[j], row[i]

    def add_col_multiple(src: int, dst: int, factor: int) -> None:
        if factor:
            for row in a:
                row[dst] += factor * row[src]
            for row in v:
                row[dst] += factor * row[src]

    rank = 0
    limit = min(m, n)
    while rank < limit:
        t = rank
        # Move the smallest nonzero entry of the trailing submatrix to (t, t).
        pivot = None
        for i in range(t, m):
            for j in range(t, n):
                if a[i][j] != 0 and (
                    pivot is None or abs(a[i][j]) < abs(a[pivot[0]][pivot[1]])
                ):
                    pivot = (i, j)
        if pivot is None:
            break
        swap_rows(t, pivot[0])
        swap_cols(t, pivot[1])

        while True:
            # Clear column t below the pivot (Euclidean reduction; every swap
            # strictly shrinks |a[t][t]|, so this terminates).
            i = t + 1
            while i < m:
                if a[i][t] != 0:
                    quotient = a[i][t] // a[t][t]
                    add_row_multiple(t, i, -quotient)
                    if a[i][t] != 0:
                        swap_rows(t, i)
                        continue
                i += 1
            # Clear row t to the right of the pivot.
            j = t + 1
            while j < n:
                if a[t][j] != 0:
                    quotient = a[t][j] // a[t][t]
                    add_col_multiple(t, j, -quotient)
                    if a[t][j] != 0:
                        swap_cols(t, j)
                        continue
                j += 1
            # Row clearing may have dirtied the column and vice versa.
            if any(a[i][t] != 0 for i in range(t + 1, m)):
                continue
            if any(a[t][j] != 0 for j in range(t + 1, n)):
                continue
            # The pivot must divide the whole trailing submatrix; otherwise
            # pull an offending entry into row t and keep reducing.
            bad = None
            for i in range(t + 1, m):
                for j in range(t + 1, n):
                    if a[i][j] % a[t][t] != 0:
                        bad = (i, j)
                        break
                if bad is not None:
                    break
            if bad is None:
                break
            add_row_multiple(bad[0], t, 1)

        if a[t][t] < 0:
            negate_row(t)
        rank += 1

    diagonal = [a[i][i] for i in range(rank)]
    return SmithDecomposition(d=a, u=u, v=v, rank=rank, diagonal=diagonal)


@dataclass
class Obstruction:
    """Canonical divisibility obstruction to integer solvability.

    Derived purely from invertible integer row transformations: the
    transformed target is ``U @ b`` and the pivot is the Smith diagonal
    entry of the same row.
    """

    kind: str  # "non_divisible" | "zero_row"
    row: int  # row index in the Smith-transformed system
    pivot: int  # Smith pivot of that row (0 for a zero row)
    transformed_target: int  # (U @ b)[row]
    remainder: int  # transformed_target mod pivot (target itself for zero rows)
    u_row: List[int]  # row of U producing the transformed target


@dataclass
class SolveResult:
    solvable: bool
    solution: List[int] = field(default_factory=list)
    obstruction: Obstruction | None = None
    smith: SmithDecomposition | None = None
    transformed_target: List[int] = field(default_factory=list)  # U @ b
    homogeneous_basis: List[List[int]] = field(default_factory=list)


def solve_diophantine(
    matrix: Sequence[Sequence[int]], target: Sequence[int]
) -> SolveResult:
    """Decide integer solvability of ``matrix @ x = target`` exactly.

    Returns a particular integer solution (free variables set to zero)
    plus a basis of the homogeneous solution space when solvable, or the
    canonical divisibility obstruction when not.
    """
    m = len(matrix)
    n = len(matrix[0]) if m else 0
    if m == 0 or n == 0:
        raise ValueError("约束矩阵不能为空")
    if len(target) != m:
        raise ValueError("目标向量长度必须与约束条数一致")
    if any(len(row) != n for row in matrix):
        raise ValueError("约束矩阵各行长度必须一致")

    smith = smith_normal_form(matrix)
    b_transformed = [
        sum(smith.u[i][k] * target[k] for k in range(m)) for i in range(m)
    ]

    for i in range(smith.rank):
        pivot = smith.diagonal[i]
        remainder = b_transformed[i] % pivot
        if remainder != 0:
            return SolveResult(
                solvable=False,
                obstruction=Obstruction(
                    kind="non_divisible",
                    row=i,
                    pivot=pivot,
                    transformed_target=b_transformed[i],
                    remainder=remainder,
                    u_row=list(smith.u[i]),
                ),
                smith=smith,
                transformed_target=b_transformed,
            )
    for i in range(smith.rank, m):
        if b_transformed[i] != 0:
            return SolveResult(
                solvable=False,
                obstruction=Obstruction(
                    kind="zero_row",
                    row=i,
                    pivot=0,
                    transformed_target=b_transformed[i],
                    remainder=b_transformed[i],
                    u_row=list(smith.u[i]),
                ),
                smith=smith,
                transformed_target=b_transformed,
            )

    y = [b_transformed[i] // smith.diagonal[i] for i in range(smith.rank)]
    y.extend([0] * (n - smith.rank))
    solution = [sum(smith.v[i][j] * y[j] for j in range(n)) for i in range(n)]

    # Exact self-check: the candidate solution must reproduce the target.
    for i in range(m):
        if sum(matrix[i][j] * solution[j] for j in range(n)) != target[i]:
            raise ArithmeticError("内部校验失败：解未复现目标向量")

    basis = [
        [smith.v[i][j] for i in range(n)] for j in range(smith.rank, n)
    ]
    return SolveResult(
        solvable=True,
        solution=solution,
        smith=smith,
        transformed_target=b_transformed,
        homogeneous_basis=basis,
    )

File: src/diophantine.js
// 任意精度丢番图方程组判定核心
// 对 A x = b（A 为 m×n 整数矩阵，b 为 m 维整数向量）：
//   1. 用扩展欧几里得算法构造 Smith 正规形 S = U A V（U、V 为幺模整数矩阵，det = ±1）；
//   2. 令 c = U b，则原方程组等价于 S y = c（x = V y）；
//   3. S 为对角矩阵 diag(d1,...,dr,0,...)，且 d1|d2|...|dr：
//      - 对每个主元 dk，必须有 dk | c_k，否则无解（规范除尽障碍）；
//      - 主元为 0 的行必须有 c_k = 0，否则无解（0·x = 非零）；
//   4. 有解时给出一组整数特解 x = V y（y_k = c_k/d_k，自由变量取 0），
//      齐次解空间由 V 的自由列给出。
// 全程使用 BigInt，不经过浮点数，商仅在确实整除时进行。

export class CancellationError extends Error {
  constructor() {
    super('复核已取消');
    this.name = 'CancellationError';
    this.code = 'ECANCELLED';
  }
}

const abs = (x) => (x < 0n ? -x : x);
const isBigInt = (x) => typeof x === 'bigint';

// 扩展欧几里得：返回 { g, u, v }，g > 0 且 u*a + v*b = g
export function egcd(a, b) {
  if (!isBigInt(a) || !isBigInt(b)) throw new TypeError('egcd 输入必须为 BigInt');
  const sa = a < 0n ? -1n : 1n;
  const sb = b < 0n ? -1n : 1n;
  let oldR = a < 0n ? -a : a;
  let r = b < 0n ? -b : b;
  let oldS = 1n;
  let s = 0n;
  let oldT = 0n;
  let t = 1n;
  while (r !== 0n) {
    const q = oldR / r; // 非负整数之间的截断除法即精确商
    [oldR, r] = [r, oldR - q * r];
    [oldS, s] = [s, oldS - q * s];
    [oldT, t] = [t, oldT - q * t];
  }
  // oldR=0 的情形（两数皆 0）在本算法调用点不会出现；这里保持 g>=0
  return { g: oldR, u: oldS * sa, v: oldT * sb };
}

function identity(size) {
  const m = new Array(size);
  for (let i = 0; i < size; i++) {
    m[i] = new Array(size).fill(0n);
    m[i][i] = 1n;
  }
  return m;
}

function cloneMatrix(A) {
  return A.map((row) => row.slice());
}

function swapRows(M, a, b) {
  const t = M[a];
  M[a] = M[b];
  M[b] = t;
}

function swapCols(M, a, b) {
  for (let i = 0; i < M.length; i++) {
    const t = M[i][a];
    M[i][a] = M[i][b];
    M[i][b] = t;
  }
}

function matmul(A, B) {
  const m = A.length;
  const k = B.length;
  const n = k === 0 ? 0 : B[0].length;
  const C = Array.from({ length: m }, () => new Array(n).fill(0n));
  for (let i = 0; i < m; i++) {
    for (let t = 0; t < k; t++) {
      const aik = A[i][t];
      if (aik === 0n) continue;
      for (let j = 0; j < n; j++) C[i][j] += aik * B[t][j];
    }
  }
  return C;
}

function matvec(A, x) {
  const m = A.length;
  const n = x.length;
  const y = new Array(m).fill(0n);
  for (let i = 0; i < m; i++) {
    let s = 0n;
    for (let j = 0; j < n; j++) s += A[i][j] * x[j];
    y[i] = s;
  }
  return y;
}

// 计算 Smith 正规形。
// hooks: { isCancelled?: () => boolean, yield?: () => Promise<void>, onStage?: (info)=>void }
// 返回 { S, U, V, pivots: BigInt[], rank, m, n, operations }
export async function smithNormalForm(Ain, hooks = {}) {
  const isCancelled = hooks.isCancelled || null;
  const yieldToLoop = hooks.yield || null;
  const onStage = hooks.onStage || null;
  let opsBudget = 0;
  const maybeYield = async (force = false) => {
    if (isCancelled && isCancelled()) throw new CancellationError();
    if (yieldToLoop && (force || ++opsBudget >= 2048)) {
      opsBudget = 0;
      await yieldToLoop();
      if (isCancelled && isCancelled()) throw new CancellationError();
    }
  };

  const m = Ain.length;
  const n = m === 0 ? 0 : Ain[0].length;
  const S = cloneMatrix(Ain);
  const U = identity(m);
  const V = identity(n);
  const operations = [];

  let k = 0;
  const limit = Math.min(m, n);
  while (k < limit) {
    await maybeYield(true);
    if (onStage) onStage({ stage: k + 1, total: limit });

    // 1. 在右下角子矩阵中选取【绝对值最小】的非零元作为主元，经行/列交换移到 (k,k)。
    //    最小主元策略是抑制朴素 Smith 算法中间系数膨胀的标准手段，
    //    也使后续 gcd 下降链具有接近欧几里得算法的步数。
    let pi = -1;
    let pj = -1;
    let pAbs = 0n;
    for (let i = k; i < m; i++) {
      for (let j = k; j < n; j++) {
        const v = S[i][j];
        if (v === 0n) continue;
        const a = v < 0n ? -v : v;
        if (pi === -1 || a < pAbs) {
          pi = i;
          pj = j;
          pAbs = a;
        }
      }
    }
    if (pi === -1) break; // 子矩阵全零，秩为 k

    if (pi !== k) {
      swapRows(S, k, pi);
      swapRows(U, k, pi);
      operations.push({ type: 'row-swap', a: k, b: pi });
    }
    if (pj !== k) {
      swapCols(S, k, pj);
      swapCols(V, k, pj);
      operations.push({ type: 'col-swap', a: k, b: pj });
    }

    // 2. 消去主元列下方 / 主元行右方元素。
    //    - 若当前主元 p 整除该元素：使用平凡 Bezout 系数（u=sign(p), v=0），
    //      只对主元行/列变号即可把该元素清零，不向另一方向引入非零；
    //    - 若不整除：扩展欧几里得给出新主元 g=gcd(p,x)，|g|<|p|，主元严格变小，
    //      整轮重来（自然数严格递减保证终止）。
    let reduced = true;
    while (reduced) {
      reduced = false;

      // 主元列（行变换，同步作用于 U）
      for (let i = k + 1; i < m; i++) {
        const x = S[i][k];
        if (x === 0n) continue;
        const p = S[k][k];
        let u, v, g;
        if (x % p === 0n) {
          g = p < 0n ? -p : p;
          u = p < 0n ? -1n : 1n;
          v = 0n;
        } else {
          ({ g, u, v } = egcd(p, x));
        }
        const rk = S[k].slice();
        const ri = S[i].slice();
        const xg = x / g;
        const pg = p / g; // g|p、g|x 保证精确
        for (let j = 0; j < n; j++) {
          S[k][j] = u * rk[j] + v * ri[j];
          S[i][j] = -xg * rk[j] + pg * ri[j];
        }
        const uk = U[k].slice();
        const ui = U[i].slice();
        for (let j = 0; j < m; j++) {
          U[k][j] = u * uk[j] + v * ui[j];
          U[i][j] = -xg * uk[j] + pg * ui[j];
        }
        operations.push({ type: 'row-combine', a: k, b: i, u: u.toString(), v: v.toString() });
        await maybeYield();
        if (g < abs(p)) {
          reduced = true; // 主元严格变小，整轮重来
          break;
        }
      }
      if (reduced) continue;

      // 主元行（列变换，同步作用于 V）
      for (let j = k + 1; j < n; j++) {
        const x = S[k][j];
        if (x === 0n) continue;
        const p = S[k][k];
        let u, v, g;
        if (x % p === 0n) {
          g = p < 0n ? -p : p;
          u = p < 0n ? -1n : 1n;
          v = 0n;
        } else {
          ({ g, u, v } = egcd(p, x));
        }
        const ck = new Array(m);
        const cj = new Array(m);
        for (let r = 0; r < m; r++) {
          ck[r] = S[r][k];
          cj[r] = S[r][j];
        }
        const vk = new Array(n);
        const vj = new Array(n);
        for (let r = 0; r < n; r++) {
          vk[r] = V[r][k];
          vj[r] = V[r][j];
        }
        const xg = x / g;
        const pg = p / g;
        for (let r = 0; r < m; r++) {
          S[r][k] = u * ck[r] + v * cj[r];
          S[r][j] = -xg * ck[r] + pg * cj[r];
        }
        for (let r = 0; r < n; r++) {
          V[r][k] = u * vk[r] + v * vj[r];
          V[r][j] = -xg * vk[r] + pg * vj[r];
        }
        operations.push({ type: 'col-combine', a: k, b: j, u: u.toString(), v: v.toString() });
        await maybeYield();
        if (abs(g) < abs(p)) {
          reduced = true;
          break;
        }
      }
      if (reduced) continue;

      // 3. 主元行/列已清零，但若主元不能整除右下角子矩阵某元素，
      //    用幺模列变换 (Ck,Cj) <- (Ck+Cj, Ck)（行列式 -1）把该元素
      //    引入主元列下沿，下一轮 gcd 必使主元严格变小。
      const p = S[k][k];
      repair: for (let i = k + 1; i < m; i++) {
        for (let j = k + 1; j < n; j++) {
          if (S[i][j] % p !== 0n) {
            const ck = new Array(m);
            const cj = new Array(m);
            for (let r = 0; r < m; r++) {
              ck[r] = S[r][k];
              cj[r] = S[r][j];
            }
            const vk = new Array(n);
            const vj = new Array(n);
            for (let r = 0; r < n; r++) {
              vk[r] = V[r][k];
              vj[r] = V[r][j];
            }
            for (let r = 0; r < m; r++) {
              S[r][k] = ck[r] + cj[r];
              S[r][j] = ck[r];
            }
            for (let r = 0; r < n; r++) {
              V[r][k] = vk[r] + vj[r];
              V[r][j] = vk[r];
            }
            operations.push({ type: 'col-repair', a: k, b: j });
            reduced = true;
            await maybeYield();
            break repair;
          }
        }
      }
    }

    // 4. 规范主元符号为正（行乘 -1，仍为可逆整数行变换）
    if (S[k][k] < 0n) {
      for (let j = 0; j < n; j++) S[k][j] = -S[k][j];
      for (let j = 0; j < m; j++) U[k][j] = -U[k][j];
      operations.push({ type: 'row-sign', a: k });
    }
    k++;
  }

  const rank = k;
  const pivots = [];
  for (let i = 0;i< rank; i++) pivots.push(S[i][i]);
  return { S, U, V, pivots, rank, m, n, operations };
}

// 判定 A x = b 的整数可解性并构造证据与特解。
export async function solveDiophantine(Ain, bin, hooks = {}) {
  const A = cloneMatrix(Ain);
  const b = bin.slice();
  const { S, U, V, pivots, rank, m, n, operations } = await smithNormalForm(A, hooks);
  const c = matvec(U, b);

  const obstructions = [];

  // 主元行：d_k 必须整除变换后的目标 c_k
  for (let i = 0; i < rank; i++) {
    const d = pivots[i];
    if (c[i] % d !== 0n) {
      const uRow = U[i];
      const uA = new Array(n).fill(0n);
      for (let t = 0; t < m; t++) {
        const coeff = uRow[t];
        if (coeff === 0n) continue;
        for (let j = 0; j < n; j++) uA[j] += coeff * A[t][j];
      }
      // 余数按非负规范给出（0 <= r < d）
      let r = c[i] % d;
      if (r < 0n) r += d;
      obstructions.push({
        type: 'divisibility',
        row: i,
        pivot: d,
        transformedTarget: c[i],
        remainder: r,
        uRow,
        uA,
      });
    }
  }
  // 零主元行：0·x 必须等于 0
  for (let i = rank; i < m; i++) {
    if (c[i] !== 0n) {
      const uRow = U[i];
      const uA = new Array(n).fill(0n);
      for (let t = 0; t < m; t++) {
        const coeff = uRow[t];
        if (coeff === 0n) continue;
        for (let j = 0; j < n; j++) uA[j] += coeff * A[t][j];
      }
      obstructions.push({
        type: 'zero-row',
        row: i,
        pivot: 0n,
        transformedTarget: c[i],
        remainder: c[i],
        uRow,
        uA,
      });
    }
  }

  const solvable = obstructions.length === 0;
  let solution = null;
  let nullspace = null;

  if (solvable) {
    // y_k = c_k/d_k，自由变量取 0；x = V y
    const y = new Array(n).fill(0n);
    for (let i = 0; i < rank; i++) y[i] = c[i] / pivots[i];
    solution = matvec(V, y);

    // 齐次解基：V 中对应自由变量 y_{rank..n-1} 的列
    nullspace = [];
    for (let j = rank; j < n; j++) {
      const vec = new Array(n);
      for (let i = 0; i < n; i++) vec[i] = V[i][j];
      nullspace.push(vec);
    }
  }

  return {
    solvable,
    A,
    b,
    S,
    U,
    V,
    c,
    pivots,
    rank,
    m,
    n,
    operations,
    obstructions,
    solution,
    nullspace,
  };
}

// 用解回代核验每条原始约束：各项乘积、左侧和、是否等于目标
export function verifyConstraints(result) {
  const { A, b, solution, variables, n, m } = result;
  const rows = [];
  for (let i = 0; i < m; i++) {
    const terms = [];
    let sum = 0n;
    for (let j = 0; j < n; j++) {
      const coeff = A[i][j];
      const value = solvableOrNull(result) ? solution[j] : null;
      const product = value === null ? null : coeff * value;
      if (product !== null) sum += product;
      terms.push({
        coefficient: coeff,
        variable: variables ? variables[j] : `x${j + 1}`,
        value,
        product,
      });
    }
    rows.push({ index: i, terms, sum, target: b[i], balanced: sum === b[i] });
  }
  return rows;
}

function solvableOrNull(result) {
  return result.solvable && result.solution;
}

export const _internal = { matmul, matvec, identity, cloneMatrix };
